docs(homeserver): Improve developer docs and readmes (#431)
pubky/pubky-core: docs/INSTALL.md

# Install and Run Pubky Homeserver

> This guide is for running a standalone Pubky homeserver. For local app development against a local testnet, see [Local Development](./LOCAL_DEVELOPMENT.md). For contributor test databases and CI setup, see [Testing](./TESTING.md).

Author: Reading this section feels bumpy. Can we do something like this?

```suggestion
This guide is for running a standalone Pubky homeserver.

> - For local app development against a local testnet, see [Local Development](./LOCAL_DEVELOPMENT.md). 
> - For contributor test databases and CI setup, see [Testing](./TESTING.md).
```

I am not even sure if my suggestion is the way to go. To put my "feeling bumpy" into words:

I ready 1 sentence what this file is about:
> This guide is for running a standalone Pubky homeserver.

And then I read two sentences that are unrelated:
>  For local app development against a local testnet, see [Local Development](./LOCAL_DEVELOPMENT.md). For contributor test databases and CI setup, see [Testing](./TESTING.md).

That confuses me. Maybe we should just remove these two sentences?

Happy to hear what you think.

Reviewer: Made into a `Looking for something else?` note

## Contents

- [Quick Start with Docker Compose](#quick-start-with-docker-compose)
- [Manual Setup](#manual-setup)
  - [Install the Homeserver](#install-the-homeserver)
    - [Release Binary](#release-binary) | [Build From Source](#build-from-source) ([Cargo](#with-cargo) | [Docker](#with-docker))
  - [Initialise the Data Directory](#initialise-the-data-directory)
  - [Set Up PostgreSQL](#set-up-postgresql)
    - [Docker](#docker-1) | [Native](#native) | [Existing](#existing-instance)
  - [Configure the Homeserver with PostgreSQL](#configure-the-homeserver-with-postgresql)
  - [Run](#run)
- [Configuration](#configuration)
- [Production Notes](#production-notes)
- [Troubleshooting](#troubleshooting)

## Quick Start with Docker Compose

> The [pubky-docker](https://github.com/pubky/pubky-docker) repository provides a Docker Compose setup that runs the homeserver and PostgreSQL together using pre-built images. This is the fastest way to get a homeserver running.

Author: pubky-docker is not intended for regular homeserver running.

<img width="1774" height="614" alt="Image" src="https://github.com/user-attachments/assets/fc637fb9-1bba-4945-9a7a-f7cb8c994c9b" />

Reviewer: added that same warning here

Author: Did you add this to INSTALL.md? This is not what I meant and feels unnecessary.

<img width="2070" height="534" alt="CleanShot 2026-07-07 at 08 30 02@2x" src="https://github.com/user-attachments/assets/99905fd2-a3b4-4047-a44b-d88c485288c7" />

What I meant is that you recommend to run the homeserver with pubky-docker but pubky-docker is explicitly not made for this. I think we should remove pubky-docker mentions in here.

Requires [Docker Engine](https://docs.docker.com/engine/install/).

```bash
git clone https://github.com/pubky/pubky-docker.git
cd pubky-docker
cp .env-sample .env

> docker compose --profile backend up -d postgres homeserver

Author: NIT: Might be worth adding `--no-build`.

AI output:

The Docker Compose quick start says this uses pre-built images, but the command omits --no-build. In pubky-docker, the homeserver service has both image and build: ../pubky-core, so on a fresh clone Compose may try to build from a missing sibling ../pubky-core instead of pulling the published image. The upstream pubky-docker README uses docker compose --profile backend up -d --no-build.

```

This starts PostgreSQL and the homeserver with persistent storage, networking, and healthchecks pre-configured. The homeserver is available at `http://localhost:6286` and the admin API at `http://localhost:6288`.

To customise the homeserver, edit `homeserver.config.toml` in the pubky-docker directory. See [Configuration](#configuration) for available settings and [Production Notes](#production-notes) before exposing the server publicly.

## Manual Setup

The steps below walk through installing the homeserver binary, setting up PostgreSQL, and configuring everything individually. Use this path if you need full control over each component.

## Install the Homeserver

### Release Binary

Download the latest non-prerelease archive from the [Pubky Core releases page](https://github.com/pubky/pubky-core/releases). Choose the archive for your operating system and CPU architecture, extract it, and place `pubky-homeserver` somewhere on your `PATH`.

Requires `curl`. On Ubuntu/Debian: `apt install curl`.

```bash
curl -LO https://github.com/pubky/pubky-core/releases/download/vx.x.x/pubky-core-vx.x.x-linux-amd64.tar.gz
tar -xf pubky-core-vx.x.x-linux-amd64.tar.gz

> cp pubky-core-vx.x.x-linux-amd64/pubky-homeserver /usr/local/bin

Author: NIT: Can we do the copy in two steps? 1. cd into dir 2. cp. Not critical but feels a bit easier to comprehend.

```suggestion
curl -LO https://github.com/pubky/pubky-core/releases/download/vx.x.x/pubky-core-vx.x.x-linux-amd64.tar.gz
tar -xf pubky-core-vx.x.x-linux-amd64.tar.gz
cd pubky-core-vx.x.x-linux-amd64
cp pubky-homeserver /usr/local/bin
```

```

> 

Author: Can we add a quick verifying section here where the user runs `pubky-homeserver --version` after the binary is copied? This would ensure that the users downloaded the correct release binaries for his architecture. This is a common mistake users.

Something like:

Run `pubky-homeserver --version`. If everything is installed correctly it should output the binary version.

### Build From Source

On Ubuntu you might need: `apt install build-essential git curl`.

Clone the repository:

```bash
git clone https://github.com/pubky/pubky-core.git
cd pubky-core
git checkout vx.x.x   # Pick a version
```

> #### With Cargo

Author: The "With Cargo" and "With Docker" titles feel weird. Shouldn't they be called "Build binaries" and "Builder Docker image" or so? Reading "With Cargo" makes me ask "Yes, I am building with Cargo. It's a Rust app. What is this about?".

Maybe it's even worth splitting this guide in two: One for Binary build and one for the docker image build. Hard to say.

<img width="1382" height="1280" alt="Image" src="https://github.com/user-attachments/assets/f587946a-fb61-4cf7-b31d-8a7c401c96ce" />

Reviewer: changed to `Build a binary with Cargo` and `Build a Docker image`

Make sure you have the Rust toolchain installed and working.

<details>
<summary>Install Rust</summary>

Quick setup using [rustup](https://rustup.rs/) (recommended) on macOS or Linux:

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
source ~/.cargo/env
```

For other platforms or methods, see the [Rust Install Guide](https://rust-lang.org/tools/install/).

</details>

Build and place the binary on your `PATH`:

```bash
cargo build --release -p pubky-homeserver
cp ./target/release/pubky-homeserver /usr/local/bin
```

#### With Docker

> Build the homeserver image using the [Dockerfile](../Dockerfile):

Author: Do you think this sentence is necessary or can it be dropped for the sake of being consice?

```suggestion
```

Reviewer: I dont think it hurts to explain what were doing before the code block, its what we do in the rest of the file for other commands.

```bash
docker build --build-arg BUILD_TARGET=homeserver -t pubky-homeserver .
```

You can run the homeserver directly in Docker:

```bash

> docker run -it --network=host pubky-homeserver

Author: How do I configure docker? You describe creating the config.toml and so on with the binary but it's missing with docker completely, isn't?

```

> Use `--network=host` so the container can reach PostgreSQL on the host and expose its endpoints.

Author: You run the container here before you setup Postgres. If a users follows from top to bottom, he will fail here. You didn't do this the "with cargo" binary build before.

<img width="2152" height="200" alt="Image" src="https://github.com/user-attachments/assets/79bf77ad-233a-439b-b37f-7aa60b20e1d0" />

Reviewer: Replaced with a `docker run --rm pubky-homeserver --version`

> **macOS note:** Docker Desktop for macOS does not support `--network=host`. Use `-p 6286:6286 -p 6287:6287 -p 6288:6288 -p 6289:6289` instead to map the ports manually.

> See the Docker documentation for volume mounts and other options.

Author: To run the docker homeserver correctly in production, one needs a volume to persist the data no? So it should be described here, shouldn't it?

Or copy the binary out of the image and place it on your `PATH`:

```bash
docker create --name tmp-hs pubky-homeserver
docker cp tmp-hs:/usr/local/bin/homeserver /usr/local/bin/pubky-homeserver
docker rm tmp-hs

> ```

Author: What is this? Sounds unnecessary? The image might not even have the same compute architecture than the host.

Reviewer: I was hoping to keep things to a single run method to make later docs simpler - ie just one `pubky-homeserver` command. This wont work though, youre right.  
Removed this option completely.

## Initialise the Data Directory

Create the data directory, default `config.toml`, and server keypair without starting the server or connecting to PostgreSQL:

```bash
pubky-homeserver init

> ```

Author: ```
pubky-homeserver init
error: unexpected argument 'init' found

Usage: pubky-homeserver [OPTIONS]

For more information, try '--help'.
```

Reviewer: Added note about requiring v0.10 and what to do if on <0.10

This creates `~/.pubky/` with a sample config and a fresh server keypair. To use a different path:

```bash
pubky-homeserver --data-dir /path/to/pubky-data init
```

## Set Up PostgreSQL

> The homeserver requires a running PostgreSQL instance with an empty database. It runs migrations automatically on startup. The default connection string is `postgres://localhost:5432/pubky_homeserver`.

Author: For the sake of being concise (the user wants to get this running asap), should we remove these two (unnecessary) sentences?
```suggestion
The homeserver requires a running PostgreSQL instance with an empty database.
```

### Docker

> Requires [Docker Engine](https://docs.docker.com/engine/install/ubuntu/).

Author: Mentioning this down here is a bit late. Could be moved up where one actually starts using docker.

Reviewer: This is the first mention of docker in the Postgres section. Copied  this line to the Docker build instructions.

Start a PostgreSQL container with the `pubky_homeserver` database:

```bash
docker run --name pubky-postgres \
  -e POSTGRES_USER=postgres \
  -e POSTGRES_PASSWORD=postgres \
  -e POSTGRES_DB=pubky_homeserver \
  -p 127.0.0.1:5432:5432 \
  -v postgres-data:/var/lib/postgresql \
  -d postgres:18
```

Verify it's running and the database exists:

```bash
docker exec pubky-postgres psql -U postgres -c '\l pubky_homeserver'

> ```

Author: Is this verification step necessary? This docker image guarantees to create `POSTGRES_DB`. So it feels like it's not needed?

```

Start the server (not needed on systems with `systemd`, where PostgreSQL starts automatically):

```bash
pg_ctlcluster $(pg_lsclusters -h | awk '{print $1, $2}') start
```

Set a password and create the database:

```bash
su - postgres -c "psql -c \"ALTER USER postgres PASSWORD 'postgres';\" && createdb pubky_homeserver"
```

Verify the connection:

```bash
psql "postgres://postgres:postgres@localhost:5432/pubky_homeserver" -c '\conninfo'
```

### Existing instance

Create a database on your existing PostgreSQL instance:

```bash
createdb -h <HOST> -U <USER> pubky_homeserver
```

Verify the connection:

```bash
psql "postgres://<USER>:<PASSWORD>@<HOST>:5432/pubky_homeserver" -c '\conninfo'
```

## Configure the Homeserver with PostgreSQL

Update `database_url` in `~/.pubky/config.toml` to match your PostgreSQL connection string. For the Docker and Native examples above:

```toml
[general]
database_url = "postgres://postgres:postgres@localhost:5432/pubky_homeserver"
```

Replace the credentials and host if using an existing instance.

## Run

Start the homeserver:

```bash
pubky-homeserver
```

From source:

```bash
cargo run -p pubky-homeserver

> ```

Author: We already built from source. To stay concise, I suggest we remove this

The default endpoints are:

| Endpoint | Default |
| --- | --- |
| Public HTTP API | `http://127.0.0.1:6286` |
| Pubky TLS API | `127.0.0.1:6287` |
| Admin API | `http://127.0.0.1:6288` |
| Metrics API | `http://127.0.0.1:6289` |

Standalone homeservers require signup tokens by default. Generate one through the admin API:

```bash
curl -X GET "http://127.0.0.1:6288/generate_signup_token" \
  -H "X-Admin-Password: admin"
```

> You can also open signup entirely for a private or temporary deployment by setting `signup_mode = "open"` in `config.toml` (see [Configuration](#configuration)).

Author: Not really necessary here wdyt? This is more of a side feature than something we should mention here. Default should be signup tokens.

If you would like to test your homeserver with example clients, see [Run Examples](./LOCAL_DEVELOPMENT.md#run-examples) in the Local Development guide.

## Configuration

Important settings in `config.toml`:

| Setting | Purpose |
| --- | --- |
| `general.database_url` | PostgreSQL connection string. |
| `general.signup_mode` | Signup policy: `token_required` or `open`. |
| `drive.icann_listen_socket` | Regular HTTP API listen address. |
| `drive.pubky_listen_socket` | Pubky TLS API listen address. |
| `storage.type` | Storage backend: `file_system`, `google_bucket`, or `in_memory`. |
| `admin.enabled` | Enables the admin API. |
| `admin.listen_socket` | Admin API listen address. |
| `metrics.enabled` | Enables Prometheus metrics. |
| `metrics.listen_socket` | Metrics API listen address. |
| `pkdns.public_ip` | Public IP advertised for Pubky discovery. |

> | `pkdns.icann_domain` | Domain used for regular browser HTTP access. |

Author: - Metrics are very much a niche feature for devops. Wouldn't feature it here as "important settings".
- signup_mode is not really important IMO
- `admin.enabled` is not so important to be featured here IMO.
- `pkdns` might be worth having a full section on. But I think you are already on this?

Reviewer: removed all apart from admin.

Review the full documented sample at [`pubky-homeserver/config.sample.toml`](../pubky-homeserver/config.sample.toml).

## Production Notes

Before using a homeserver in production:

> - Use a persistent PostgreSQL instance with password authentication and back it up. Do not use trust auth in production.

Author: Why not?

Reviewer: Removed. This is a relic from a previous version in which I attempted to simplify things by not setting a user and password on the postgres server. Decided against that change - all docs set user and pass by default now so this no longer needs saying.

- Back up the homeserver `secret` file and any filesystem or bucket storage.
- Do not expose the admin or metrics APIs to the public internet.
- Change the default admin password in `[admin].admin_password`.
- Configure `pkdns.public_ip`, `pkdns.icann_domain`, and public ports for your deployment.

> - Put the regular HTTP API behind a reverse proxy if you need browser-compatible HTTPS.

Author: Worth having this better described, what do you think? We have two sockets (pubky tls + regular http) so it's not obvious how to do it, is it?

Reviewer: Improved this note and also will link to full pubky.org guides when they exist

- Use persistent filesystem storage or a configured bucket backend, not in-memory storage.

> - Monitor logs, PostgreSQL health, disk usage, and storage backend errors.

Author: ```suggestion
```

## Troubleshooting

### `database "pubky_homeserver" does not exist`

Create the database:

```bash
createdb -h <HOST> -U <USER> pubky_homeserver
```

> Or update `[general].database_url` in `~/.pubky/config.toml` to point at an existing database.

Author: Is this for the docker or fixed install case? I assume fixed install. It's not obvious when reading though.

Reviewer: Added both install method commands to each troubleshooting item. Im getting close to creating a separate doc for postgres install, or seperate docs for each of the install methods. Lets see how it grows

### PostgreSQL Connection Refused

Make sure PostgreSQL is running and listening on the host and port in `general.database_url`.

For the Docker examples above, check the container:

```bash
docker ps --filter name=pubky-postgres

> ```

Author: Does this command answer if PG is running and listening and the correct host/port?

Reviewer: Improved this section generally

# Install and Run Pubky Homeserver

> How to set up and operate a Pubky homeserver.

Author: I think we need to say somewhere that this guide is intended for Ubuntu 26. Otherwise people will be confused if they want to install it on Windows. What do you think?

Reviewer: i think that is best for now.

Limiting guide to Debian-based linux for simplicity. Can add other guides later if they are called for.

> **Looking for something else?**
> See [Pubky Testnet](../pubky-testnet/README.md) for running a local development testnet and [Testing](./TESTING.md) for test databases and CI setup.

## Contents

- [Install the Homeserver](#install-the-homeserver)
    - [Release Binary](#release-binary) | [Build From Source](#build-from-source) ([Cargo](#build-a-binary-with-cargo) | [Docker](#build-a-docker-image))
  - [Initialise the Data Directory](#initialise-the-data-directory)
  - [Set Up PostgreSQL](#set-up-postgresql)
    - [Docker](#docker-1) | [Native](#native) | [Existing](#existing-instance)
  - [Configure the Homeserver with PostgreSQL](#configure-the-homeserver-with-postgresql)
  - [Run](#run)
- [Configuration](#configuration)
- [Production Notes](#production-notes)
- [Troubleshooting](#troubleshooting)

## Install the Homeserver

Pick a version from the [Pubky Core releases page](https://github.com/pubky/pubky-core/releases). The commands below use this variable, so set it first:

```bash

> PUBKY_CORE_VERSION=0.9.3

Author: AI: The guide selects version 0.9.3, then instructs users to run pubky-homeserver init. That command only exists in v0.10+, so the primary installation path fails.

Reviewer: Hmmm true. Let me leave as is for now with a note to myself to update the suggested version to `0.10` when we release.

Reviewer: I decided to also remove the version set here to force the user to choose one themselves.

```

### Release Binary

Download and extract the archive (requires `curl`; on Ubuntu/Debian: `apt install curl`):

```bash
curl -LO https://github.com/pubky/pubky-core/releases/download/v${PUBKY_CORE_VERSION}/pubky-core-v${PUBKY_CORE_VERSION}-linux-amd64.tar.gz
tar -xf pubky-core-v${PUBKY_CORE_VERSION}-linux-amd64.tar.gz

> ```

Author: The download url also defines the cpu architecture like `amd64`. Should this also be extracted into a variable similiar to the `PUBKY_CORE_VERSION`?

Place the binary on your `PATH`:

```bash
cp pubky-core-v${PUBKY_CORE_VERSION}-linux-amd64/pubky-homeserver /usr/local/bin
```

Verify the install:

```bash
pubky-homeserver --version
```

### Build From Source

Install build dependencies (Ubuntu/Debian):

```bash

> apt update && apt install -y build-essential pkg-config libssl-dev git curl

Author: Should we add `sudo` to the commands?

AI review:
docs/INSTALL.md:32-42,53-56,158-173: Ubuntu/Debian commands require root access but omit sudo, including apt, copying into /usr/local/bin, pg_ctlcluster, and su - postgres. The latter commonly fails on Ubuntu because the system account is locked.

```

### Existing instance

Create a database on your existing PostgreSQL instance:

```bash
createdb -h <HOST> -U <USER> pubky_homeserver
```

Verify the connection:

```bash
psql "postgres://<USER>:<PASSWORD>@<HOST>:5432/pubky_homeserver" -c '\conninfo'
```

## Configure the Homeserver with PostgreSQL

Uncomment and set `database_url` in `~/.pubky/config.toml`. For the Docker and Native example setup above, it should look like:

```toml
[general]
database_url = "postgres://postgres:postgres@localhost:5432/pubky_homeserver"
```

Here's a handy sed command to edit as above:

```bash
sed -i 's|^# \[general\]|[general]|' ~/.pubky/config.toml
sed -i 's|^# database_url = .*|database_url = "postgres://postgres:postgres@localhost:5432/pubky_homeserver"|' ~/.pubky/config.toml
```

## Run

Start the homeserver:

```bash
pubky-homeserver
```

With Docker:

```bash
docker run -it --network=host -v ~/.pubky:/root/.pubky pubky-homeserver homeserver
```

Use `--network=host` so the container can reach PostgreSQL on the host and expose its endpoints. The volume mount shares the data directory (config and keypair) with the container.

> > **macOS note:** Docker Desktop for macOS does not support `--network=host`. Use `-p 6286:6286 -p 6287:6287 -p 6288:6288 -p 6289:6289` instead to map the ports manually.

Author: AI review, no idea if valid. Posting here because I am not sure if you ever run it on a Mac.

docs/INSTALL.md:223-228: The macOS Docker alternative is not functional. Published ports cannot reach services bound to 127.0.0.1 inside the container, and localhost in database_url points to the container rather than the host. The guide needs either Docker Desktop host networking instructions or a complete bridge setup using host.docker.internal and 0.0.0.0 container listeners.

Reviewer: Limiting guide to Debian-based linux for simplicity. Can add other guides later if they are called for

The default endpoints are:

| Endpoint | Default |
| --- | --- |
| Public HTTP API | `http://127.0.0.1:6286` |
| Pubky TLS API | `127.0.0.1:6287` |
| Admin API | `http://127.0.0.1:6288` |

> | Metrics API | `http://127.0.0.1:6289` |

Author: AI review: The metrics endpoint is listed as a default endpoint, but metrics are disabled by default.

Should we remove metrics from this guide? It's not really that important.

Standalone homeservers require signup tokens by default. Generate one through the admin API:

```bash
curl -X GET "http://127.0.0.1:6288/generate_signup_token" \
  -H "X-Admin-Password: admin"
```

## Configuration

Important settings in `config.toml`:

| Setting | Purpose |
| --- | --- |
| `general.database_url` | PostgreSQL connection string. |
| `drive.icann_listen_socket` | Regular HTTP API listen address. |
| `drive.pubky_listen_socket` | Pubky TLS API listen address. |
| `storage.type` | Storage backend: `file_system`, `google_bucket`, or `in_memory`. |
| `admin.enabled` | Enables the admin API. |
| `admin.listen_socket` | Admin API listen address. |

Review the full documented sample at [`pubky-homeserver/config.sample.toml`](../pubky-homeserver/config.sample.toml).

## Production Notes

Before using a homeserver in production:

> - Back up homeserver's state:

Author: ```suggestion
- Back up the homeserver's state:
```

  - The keypair `.pubky/secret`
  - Any user data depending on the configured option. For example, by default files are saved in `.pubky/data/files`
  - Postgres
- Do not expose the admin or metrics APIs to the public internet.
- Change the default admin password in `[admin].admin_password`.
- Configure `pkdns.public_ip`, `pkdns.icann_domain`, and public ports for your deployment.

> - The homeserver exposes two sockets: a **Pubky TLS** socket (`pubky_listen_socket`, default port 6287) and a regular **HTTP** socket (`icann_listen_socket`, default port 6286). Pubky TLS uses PKARR-based TLS and does not need a certificate so can be exposed directly. The HTTP socket serves browsers and should be put behind a reverse proxy if you need standard HTTPS with a domain certificate.

Author: ```suggestion
- The homeserver exposes two sockets: a **Pubky TLS** socket (`pubky_listen_socket`, default port 6287) and a regular **HTTP** socket (`icann_listen_socket`, default port 6286). Pubky TLS uses PKARR-based TLS and does not need a certificate, so can be exposed directly. The HTTP socket serves browsers and should be put behind a reverse proxy if you need standard HTTPS with a domain certificate.
```

- [Troubleshooting](#troubleshooting)

## Install the Homeserver

Pick a version from the [Pubky Core releases page](https://github.com/pubky/pubky-core/releases). The commands below use this variable, so set it first:

```bash
PUBKY_CORE_VERSION=0.9.3
```

### Release Binary

Download and extract the archive (requires `curl`; on Ubuntu/Debian: `apt install curl`):

```bash
curl -LO https://github.com/pubky/pubky-core/releases/download/v${PUBKY_CORE_VERSION}/pubky-core-v${PUBKY_CORE_VERSION}-linux-amd64.tar.gz
tar -xf pubky-core-v${PUBKY_CORE_VERSION}-linux-amd64.tar.gz
```

Place the binary on your `PATH`:

```bash
cp pubky-core-v${PUBKY_CORE_VERSION}-linux-amd64/pubky-homeserver /usr/local/bin
```

Verify the install:

```bash
pubky-homeserver --version
```

### Build From Source

Install build dependencies (Ubuntu/Debian):

```bash
apt update && apt install -y build-essential pkg-config libssl-dev git curl
```

Clone the repository:

```bash
git clone https://github.com/pubky/pubky-core.git
cd pubky-core
git checkout v${PUBKY_CORE_VERSION}
```

#### Build a binary with Cargo

Make sure you have the Rust toolchain installed and working.

<details>

> <summary>Install Rust</summary>

Author: NIT

```suggestion
<summary>How to Install the Rust Toolchain</summary>
```

Author: Maybe this is not good either. Maybe we should just link to the install page and avoid lengthy explanations in this regards. The docs are missing clear focus IMO (partially an initial mistake by me) so removing the section and just linking to the rustup page might be the best move.

Quick setup using [rustup](https://rustup.rs/) (recommended) on macOS or Linux:

```bash
curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
source ~/.cargo/env
```

For other platforms or methods, see the [Rust Install Guide](https://rust-lang.org/tools/install/).

</details>

Build and place the binary on your `PATH`:

```bash
cargo build --release -p pubky-homeserver
cp ./target/release/pubky-homeserver /usr/local/bin
```

Verify the install:

```bash
pubky-homeserver --version
```

#### Build a Docker image

Requires [Docker Engine](https://docs.docker.com/engine/install/).

Build the homeserver image using the [Dockerfile](../Dockerfile):

```bash
docker build --build-arg BUILD_TARGET=homeserver -t pubky-homeserver .
```

Verify the image built correctly:

```bash
docker run --rm pubky-homeserver homeserver --version

> ```

Author: Interesting, maybe out of scope of this PR but can't you make the homeserver the default executable? Reading this now, `pubky-homeserver homeserver` feels duplicate.

Edit: https://github.com/pubky/pubky-core/pull/483

The default endpoints are:

| Endpoint | Default |
| --- | --- |
| Public HTTP API | `http://127.0.0.1:6286` |
| Pubky TLS API | `127.0.0.1:6287` |
| Admin API | `http://127.0.0.1:6288` |
| Metrics API | `http://127.0.0.1:6289` |

Standalone homeservers require signup tokens by default. Generate one through the admin API:

```bash
curl -X GET "http://127.0.0.1:6288/generate_signup_token" \
  -H "X-Admin-Password: admin"
```

## Configuration

Important settings in `config.toml`:

| Setting | Purpose |
| --- | --- |
| `general.database_url` | PostgreSQL connection string. |
| `drive.icann_listen_socket` | Regular HTTP API listen address. |
| `drive.pubky_listen_socket` | Pubky TLS API listen address. |
| `storage.type` | Storage backend: `file_system`, `google_bucket`, or `in_memory`. |
| `admin.enabled` | Enables the admin API. |
| `admin.listen_socket` | Admin API listen address. |

Review the full documented sample at [`pubky-homeserver/config.sample.toml`](../pubky-homeserver/config.sample.toml).

## Production Notes

Before using a homeserver in production:

> - Use a persistent PostgreSQL instance with password authentication and back it up.

Author: Hmmm, not sure we need to say this. Additional, recommending password auth is wrong as there are many good PG auth options and passwords is not even the strongest. I would fully remove this line.

> - Back up the homeserver `secret` file and any filesystem or bucket storage.

Author: That's a good one. Maybe we can do something like this? Maybe same same but different but this feels better

```suggestion
- Back up the homeserver
  - The keypair `.pubky/secret`
  - Any user data dending on the configured option. For example, by default files are saved in `.pubky/data/files`
  - Postgres
```

- Do not expose the admin or metrics APIs to the public internet.
- Change the default admin password in `[admin].admin_password`.

> - Configure `pkdns.public_ip`, `pkdns.icann_domain`, and public ports for your deployment.

Author: I think we still need a homeserver pkdns guide, no?

> - Use persistent filesystem storage or a configured bucket backend, not in-memory storage.

Author: I think we can remove this line. By default, the filesystem is configured and they don't need to think about this. Using buckets or in-memory storage is advanced anyway.

```suggestion
```

> - The homeserver exposes two sockets: a **Pubky TLS** socket (`pubky_listen_socket`, default port 6287) and a regular **HTTP** socket (`icann_listen_socket`, default port 6286). Pubky TLS uses PKARR-based TLS and does not need a certificate so can be exposed directly. The HTTP socket serves browsers and should be put behind a reverse proxy if you need standard HTTPS with a domain certificate.

Author: Good you mention this but this should probably it's own pkdns/expose to the internet guide, don't you think?

> ## Troubleshooting

Author: I am not sure we should provide so much PG troubleshooting. It's feels out of scope. At the same time, PG connections can fail for so many reasons, it's hard to cover them all.
